drivers/serial/uart_native_pty.c

```diff
@@ -209,6 +248,144 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
 	}
 }
 
+#ifdef CONFIG_UART_ASYNC_API
+
+static int np_uart_callback_set(const struct device *dev, uart_callback_t callback, void *user_data)
+{
+	struct native_pty_status *data = dev->data;
+
+	data->async.user_callback = callback;
+	data->async.user_data = user_data;
+
+	return 0;
+}
+
+static void np_uart_tx_done_work(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct native_pty_status *data =
+		CONTAINER_OF(dwork, struct native_pty_status, async.tx_done);
+	struct uart_event evt;
+	unsigned int key = irq_lock();
+	long ret;
+
+	evt.type = UART_TX_DONE;
+	evt.data.tx.buf = data->async.tx_buf;
+	evt.data.tx.len = data->async.tx_len;
+
+	ret = nsi_host_write(data->out_fd, evt.data.tx.buf, evt.data.tx.len);
+	(void)ret;
+
+	data->async.tx_buf = NULL;
+
+	data->async.user_callback(data->async.dev, &evt, data->async.user_data);
+	irq_unlock(key);
+}
+
+static int np_uart_tx(const struct device *dev, const uint8_t *buf, size_t len, int32_t timeout)
+{
+	struct native_pty_status *data = dev->data;
+
+	if (data->async.tx_buf) {
+		/* Port is busy */
+		return -EBUSY;
+	}
+	data->async.tx_buf = buf;
+	data->async.tx_len = len;
+
+	/* Run the callback on the next tick to give the caller time to use the return value */
+	k_work_reschedule(&data->async.tx_done, K_TICKS(1));
+	return 0;
+}
+
+static int np_uart_tx_abort(const struct device *dev)
+{
+	struct native_pty_status *data = dev->data;
+	struct k_work_sync sync;
+
+	/* Cancel the callback */
+	return k_work_cancel_delayable_sync(&data->async.tx_done, &sync) ? 0 : -EFAULT;
```

Comment:
I do not know the details of this API, but does the description https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/uart.h#L800
mean that:

* a callback with tx.len = 0 should happen here?

* should this call check for nothing being pended and return -EFAULT  if so?

Reply:
Callback added, the `k_work_cancel_delayable_sync` returns `false` if the work item has not been submitted, and will therefore return `-EFAULT`.